src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/StackItem.cs

```diff
@@ -4,56 +4,66 @@
 
 using System;
 using Internal.IL;
+using Internal.TypeSystem;
 
 namespace Internal.Runtime.Interpreter
 {
     internal abstract class StackItem
```

Comment:
Allocating an object for every StackItem will have a really bad performance. I think the `StackItem` will need to be a struct with a union of possible values. It would be fix this sooner rather than later because of it will affect all code that is using the StackItem and depends on the inheritance hierarchy.

Reply:
E.g. this is what the stack item looks like in Mono:

https://github.com/mono/mono/blob/f52f29e414ce647916376b346661baacd083b5de/mono/mini/interp/interp-internals.h#L50

Reply:
> However, C# doesn't have support for union types

No but you can simulate those by using explicit struct layout. Granted, that will only work for primitive types but not for reference types, those will need to use a separate field.

Reply:
Hmm, attempting to build an IL interpreter in C# might require some fancy acrobatics if you want it to be efficient. And you may want to start with the hard stuff, not the easy stuff. That would include dealing with struct types and IL instructions such as `ldobj` and `stobj`. Even with the current `StackItem<T>` that won't be easy because it requires you to instantiate the stack item using reflection... in an AOT environment. Mono's approach of pushing pointers rather than the actual value won't be easy either, it's either a byref (and that can't be stored on the heap) or something involving `GCHandle`.

Reply:
> No but you can simulate those by using explicit struct layout

I no next to nothing about struct layouts, perhaps you can show me an example?

Reply:
It can be something like this:

```csharp
[StructLayout(LayoutKind.Explicit)]
struct StackItem
{
    [FieldOffset(0)]
    private StackValueKind _kind;

    [FieldOffset(8)]
    private int _int32;
    [FieldOffset(8)]
    private int _int64;
    [FieldOffset(8)]
    private IntPtr _nativeInt;
    [FieldOffset(8)]
    private float _float;
    [FieldOffset(8)]
    private float _double;
    [FieldOffset(8)]
    private void* _ptr;

    [FieldOffset(16)]
    private object _objref;

    // This will require special support in the runtime to work well with the GC
    // There is `RunFunctionWithConservativelyReportedBuffer` that should be pretty close. It will need
    // refactoring to make it work well for the interpretter.
    // [FieldOffset(16)]
    // private void ref _byref;
}
```